scalyr_agent/metrics/functions.py

```diff
@@ -216,41 +287,122 @@ def calculate(cls, monitor, metric_name, metric_value, timestamp=None):
             )
 
         monitor._logger.debug(
-            'Calculated rate "%s" for metric "%s" (previous_collection_timestamp=%s,previous_metric_value=%s,current_timestamp=%s,current_value=%s)'
+            'Calculated rate "%s" for metric "%s" (previous_collection_timestamp=%s,previous_metric_value=%s,current_timestamp=%s,current_value=%s,extra_fields=%s)'
             % (
                 rate_value,
                 metric_name,
                 previous_collection_timestamp,
                 previous_metric_value,
                 timestamp_s,
                 metric_value,
+                str(extra_fields),
             ),
         )
 
         rate_metric_name = "%s%s" % (metric_name, cls.METRIC_SUFFIX)
         # TODO: Use dataclass once we only support Python 3
         result = [(rate_metric_name, rate_value)]
+
+        # Periodically print cache size and function timing information
+        LOG.info(
+            "agent_monitor_rate_metric_calculation_values_cache_stats cache_entries=%s,cache_size_bytes=%s",
+            cls.LAZY_PRINT_CACHE_SIZE_LENGTH,
+            cls.LAZY_PRINT_CACHE_SIZE_BYTES,
+            limit_key="mon-met-rate-cache-stats",
+            limit_once_per_x_secs=FUNCTION_STATS_LOG_INTERVAL_SECONDS,
+        )
+
+        LOG.info(
+            "agent_rate_func_calculate_timing_stats avg=%s,min=%s,max=%s",
+            cls.LAZY_PRINT_TIMING_MIN,
+            cls.LAZY_PRINT_TIMING_MAX,
+            cls.LAZY_PRINT_TIMING_AVG,
+            limit_key="mon-rate-calc-timing-stats",
+            limit_once_per_x_secs=FUNCTION_STATS_LOG_INTERVAL_SECONDS,
+        )
+
         return result
 
     @classmethod
-    def should_calculate(cls, monitor, metric_name):
+    def should_calculate_for_monitor_and_metric(cls, monitor, metric_name):
         """
-        Return True if client side rate should be calculated for the provided metric.
+        Return True if rate should be calculated for the provided monitor and metric name.
+
+        This function is used for internal (monitor_name, metric_name) cache and additional filtering
+        which takes extra_field values into account is performed when "calculate()" method is called.
         """
         if not monitor or not monitor._global_config:
             return False
 
+        config_calculate_rate_metric_names = (
+            monitor._global_config.calculate_rate_metric_names
+        )
+        monitor_calculate_rate_metric_names = monitor.get_calculate_rate_metric_names()
+
+        # Partial entry key without extra fields suffix
         config_entry_key = "%s:%s" % (monitor.monitor_module_name, metric_name)
 
+        for config_value in chain(
+            config_calculate_rate_metric_names, monitor_calculate_rate_metric_names
```

Comment:
Right now we don't do any such checks in the config verification.

Even if user specified the same metric name multiple times in different places, that would result in no negative impact or change of the behavior since "any" check is used (aka value will be included in case the metric name is present in the config option).

Having said that, adding ``set()`` there wouldn't hurt either. Well, I guess technically it could have a potentially very tiny negative overhead since right now ``chain()`` is used (which, iirc, is a lazy generator) and we exit the for loop on first successful match. and ``set()`` would exhaust the generator and cause the complete list to be evaluated.

But yeah, that's not really important since the overhead would be so tiny and negligible that it's not even worth bring it up (unless those lists could have 100k's of values, but then we would have different issues :)).

Reply:
Yeah, didn't take into account ability to break the loop earlier. Thanks.